### PR TITLE
Adjust farmer service SSL tests for increased reliability

### PR DESCRIPTION
**Pull request by emlowe**
Currently, the SSL tests specifically one test for the farmer is flaky - this seems to have been recently introduced possibly due to changes in aiohttp, but the exact reasons for flakiness are unknown

The end goal of the test is to make sure the server rejects the SSL connection - it seems that due to aiohttp internals, SSL errors are typically silent on the server and the client may not get a consistent error for why the connection was dropped. (see https://github.com/aio-libs/aiohttp/discussions/7325)

Adjust the flaky test as follows:
Using asyncio `debug` mode and setting the test to capture the log, we can look for the SSL certificate failure in the log files as it is output when asyncio is in debug mode. This is further complicated in that on some platforms, when asyncio is in debug mode, an exception is raised that is not catchable by pytest and fails the test - so a new exception handler is used to ignore the SSL exception, as we are only testing the log output.

This is currently very stable in CI. This replaces the fix in https://github.com/Chia-Network/chia-blockchain/pull/17899

I think the value of this particular test is debatable; I currently don't think there is any way to misconfigure the code such that this test would pass - it would have to be some error in SSL, or if all SSL cert verification was turned off.

### CODE REVIEW

File: chia/_tests/core/ssl/test_ssl.py [hunk 1]
```diff
@@ -37,6 +42,41 @@ async def establish_connection(server: ChiaServer, self_hostname: str, ssl_conte
...
+        current_loop.set_exception_handler(old_handler)
+
+
+@contextlib.contextmanager
+def set_asyncio_debug():
+    asyncio.get_event_loop().set_debug(True)
+    try:
+        yield
+    finally:
+        asyncio.get_event_loop().set_debug(False)
```
Comment (altendky) [inline chia/_tests/core/ssl/test_ssl.py:77]: Let's not get confused when we end up nesting these at some point.

```suggestion
    loop = asyncio.get_event_loop()
    original_state = loop.get_debug()
    loop.set_debug(True)
    try:
        yield
    finally:
        loop.set_debug(original_state)
```

File: chia/_tests/core/ssl/test_ssl.py [hunk 2]
```diff
@@ -37,6 +42,41 @@ async def establish_connection(server: ChiaServer, self_hostname: str, ssl_conte
...
+# This is needed on linux and mac when running in asyncio debug otherwise
+# The SSL Error causes the test to exit and fail prematurely in an uncatchable way
+# This doesn't seem to work on Windows as the exception handler is never called
+# at least for this SSL failure
+#
+@contextlib.contextmanager
+def ignore_ssl_cert_error():
+    current_loop = asyncio.get_event_loop()
+    old_handler = current_loop.get_exception_handler()
+    old_handler_fn = old_handler  # or lambda _loop, ctx: loop.default_exception_handler(ctx)
```
Comment (altendky) [inline chia/_tests/core/ssl/test_ssl.py:55]: It doesn't seem like we need both of these names.

File: chia/_tests/core/ssl/test_ssl.py [hunk 3]
```diff
@@ -65,19 +104,62 @@ async def test_farmer(self, farmer_one_harvester, self_hostname):
...
+                    caplog.clear()
+                    with ignore_ssl_cert_error(), caplog.at_level(logging.DEBUG, logger="asyncio"):
+                        await establish_connection(farmer_server, self_hostname, ssl_context)
+                except Exception:
+                    pass  # ignore any exceptions and just check the expected log output below
+                finally:
+                    assert (
+                        "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate signature failure"
+                        in caplog.text
+                    )
```
Comment (altendky) [inline chia/_tests/core/ssl/test_ssl.py:162]:
```suggestion

                assert (
                    "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate signature failure"
                    in caplog.text
                )
```

File: chia/_tests/core/ssl/test_ssl.py [hunk 4]
```diff
@@ -65,19 +104,62 @@ async def test_farmer(self, farmer_one_harvester, self_hostname):
...
+                pub_key = farmer_server.root_path / "non_valid.key"
+                generate_ca_signed_cert(chia_ca_crt_path.read_bytes(), chia_ca_key_path.read_bytes(), pub_crt, pub_key)
+
+                ssl_context = ssl_context_for_client(ca_private_crt_path, ca_private_key_path, pub_crt, pub_key)
+                try:
+                    caplog.clear()
+                    with ignore_ssl_cert_error(), caplog.at_level(logging.DEBUG, logger="asyncio"):
+                        await establish_connection(farmer_server, self_hostname, ssl_context)
+                except Exception:
+                    pass  # ignore any exceptions and just check the expected log output below
```
Comment (altendky) [inline chia/_tests/core/ssl/test_ssl.py:157]: I think we want some failure here, even if we don't maintain a list of the possible exceptions.

```suggestion
                caplog.clear()
                with ignore_ssl_cert_error(), caplog.at_level(logging.DEBUG, logger="asyncio"):
                    with pytest.raises():
                        await establish_connection(farmer_server, self_hostname, ssl_context)
```

**Review by altendky — CHANGES_REQUESTED** [review]

File: chia/_tests/core/ssl/test_ssl.py (hunk 1 above)
Comment (emlowe) [reply chia/_tests/core/ssl/test_ssl.py]: yes, changed per suggestion

File: chia/_tests/core/ssl/test_ssl.py (hunk 2 above)
Comment (emlowe) [reply chia/_tests/core/ssl/test_ssl.py]: changed

File: chia/_tests/core/ssl/test_ssl.py (hunk 3 above)
Comment (emlowe) [reply chia/_tests/core/ssl/test_ssl.py]: changed

File: chia/_tests/core/ssl/test_ssl.py (hunk 4 above)
Comment (emlowe) [reply chia/_tests/core/ssl/test_ssl.py]: changed per suggestion

**Review by altendky — APPROVED** [review]

**Review by arvidn — APPROVED** [review]

**Pull request merged**